### PR TITLE
Allow SSM Parameters to specify a format using tags

### PR DESCRIPTION
**Pull request by ConnorWidtfeldt**
## Changes proposed by this PR

Currently, there is no way to access fields from vars sourced via SSM. This is because Concourse has no understanding of the format the parameter value is in. This change allows Concourse to read SSM parameter tags (with ssm:ListTagsForResource) and check for a tag called "format" which will behave the exact same as the load_vars format. Thus allowing JSON and YAML formats to be loaded. 

* [x] implement unmarshalling using a value set in the tags
* [ ] create tests for various formats, and failures in retrieving tags from AWS
* [ ] check for any remaining steps I could have missed in this...

## Notes to reviewer

Most of the code is a modified version of the load_vars handler. Maybe these should be moved into a shared location?

## Release Note

* The AWS SSM credential manager can now load different file formats when a format is specified in a tag named `format` on the parameter

### CODE REVIEW

**Comment by taylorsilva** [comment]
You may not be able to see the logs from the failed tests but they're all under the SSM test suite. You should be able to replicate the failures by running: `go test atc/creds/ssm/`

**Comment by ConnorWidtfeldt** [comment]
> You may not be able to see the logs from the failed tests but they're all under the SSM test suite. You should be able to replicate the failures by running: `go test atc/creds/ssm/`

Is there anything special I need to do to see the output of the unit tests on ci.concourse-ci.org? I'll run them locally too, but I figured I would check in case it was an easy add.

**Comment by taylorsilva** [comment]
Add yourself as a contributor in this repo: https://github.com/concourse/governance

**Comment by taylorsilva** [comment]
The downgrade test is failing. That's in `./integration/ops`.
The integration tests are slower. They bring up concourse locally using docker compose to test stuff at a higher level.

**Comment by ConnorWidtfeldt** [comment]
I should probably try running all the tests locally to see what's up. I'm surprised to see it run slower, it should only process a parameter if it's not a raw value, which is the only thing I can think of that would cause slowdowns outside of the secondary request to AWS to get the tags (these unit tests have no ways of testing for network latency, right?) 

Glad I started with a simpler PR for my first contribution to Concourse. I'm learning a ton about the setup.

File: atc/creds/ssm/ssm.go [hunk 1]
```diff
@@ -99,5 +112,63 @@ func (s *Ssm) getParameterByPath(path string) (interface{}, *time.Time, bool, er
...
+		ResourceType: aws.String(ssm.ResourceTypeForTaggingParameter),
+		ResourceId:   id,
+	})
+
+	if err == nil {
+		for i := range tags.TagList {
+			tag := strings.ToLower(*tags.TagList[i].Key)
+			if tag == "format" {
+				format = strings.ToLower(*tags.TagList[i].Value)
+			}
```
Comment (taylorsilva) [inline atc/creds/ssm/ssm.go:140]: Once you enter this case you can `break` out of the `for` loop

File: atc/creds/ssm/ssm.go [hunk 2]
```diff
@@ -99,5 +112,63 @@ func (s *Ssm) getParameterByPath(path string) (interface{}, *time.Time, bool, er
...
+		}
+	case "yml", "yaml":
+		err := yaml.Unmarshal([]byte(input), &value)
+		if err != nil {
+			return nil, err
+		}
+	case "trim":
+		value = strings.TrimSpace(string(input))
+	case "raw":
+		value = string(input)
```
Comment (taylorsilva) [inline atc/creds/ssm/ssm.go:168]:
```suggestion
		value = input
```
Redundant type casting

File: atc/creds/ssm/ssm.go [hunk 3]
```diff
@@ -99,5 +112,63 @@ func (s *Ssm) getParameterByPath(path string) (interface{}, *time.Time, bool, er
...
+	default:
+		return nil, fmt.Errorf("unknown format %s, should never happen, ", format)
```
Comment (taylorsilva) [inline atc/creds/ssm/ssm.go:170]:
```suggestion
		return nil, fmt.Errorf("unknown format %w, should never happen, ", format)
```
Use `%w`: https://pkg.go.dev/fmt#Errorf

**Review by taylorsilva — CHANGES_REQUESTED** [review]
Didn't look over tests yet but made some comments for small improvements. Mostly nits to make things more inline with the rest of the code base. This is looking great so far though! 🎉

File: atc/creds/ssm/ssm.go [hunk 4]
```diff
@@ -99,5 +112,63 @@ func (s *Ssm) getParameterByPath(path string) (interface{}, *time.Time, bool, er
...
+			tag := strings.ToLower(*tags.TagList[i].Key)
+			if tag == "format" {
+				format = strings.ToLower(*tags.TagList[i].Value)
+			}
+		}
+
+		if !isValidFormat(format) {
+			return "nil", fmt.Errorf("invalid format %s", format)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (ConnorWidtfeldt) [reply atc/creds/ssm/ssm.go:146]: Oh, that looks way better. I like to remove as much nesting as possible, and you make a good point about it correctly handling the error case.

**Pull request closed without merging**